Refresh.GameServer/Endpoints/Api/AuthenticationApiEndpoints.cs

```diff
@@ -0,0 +1,134 @@
+
+public class AuthenticationApiEndpoints : EndpointGroup
+{
+    // How many rounds to do for password hashing (BCrypt)
+    // 14 is ~1 second for logins and reset, which is fair because logins are a one-time thing
+    // 200 OK on POST '/api/v2/resetPassword' (1058ms)
+    // 200 OK on POST '/api/v2/auth' (1087ms)
+    //
+    // If increased, passwords will automatically be rehashed at login time to use the new WorkFactor
+    // If decreased, passwords will stay at higher WorkFactor until reset
+    private const int WorkFactor = 14;
+
+    [ApiEndpoint("auth", Method.Post)]
+    [Authentication(false)]
+    public Response Authenticate(RequestContext context, RealmDatabaseContext database, ApiAuthenticationRequest body)
+    {
+        GameUser? user = database.GetUserByUsername(body.Username);
+        if (user == null)
+        {
+            return new Response(new ApiErrorResponse("The username or password was incorrect."), ContentType.Json, HttpStatusCode.Forbidden);
+        }
+        
+        // database.SetUserPassword(user, null);
+        // return new Response("");
+
+        // if this is a legacy user, have them create a password on login
+        if (user.PasswordBcrypt == null)
+        {
+            ResetToken resetToken = database.GenerateResetTokenForUser(user);
+
+            ApiResetPasswordResponse resetResp = new()
+            {
+                Reason = "The account you are trying to sign into is a legacy account. Please set a password.",
+                ResetToken = resetToken.TokenData,
+            };
+
+            return new Response(resetResp, ContentType.Json, HttpStatusCode.Unauthorized);
+        }
+
+        if (BC.PasswordNeedsRehash(user.PasswordBcrypt, WorkFactor))
+        {
+            database.SetUserPassword(user, BC.HashPassword(body.PasswordSha512, WorkFactor));
+        }
+
+        if (!BC.Verify(body.PasswordSha512, user.PasswordBcrypt))
+        {
+            return new Response(new ApiErrorResponse("The username or password was incorrect."), ContentType.Json, HttpStatusCode.Forbidden);
+        }
+
+        Token token = database.GenerateTokenForUser(user, TokenType.Api);
+
+        ApiAuthenticationResponse resp = new()
+        {
+            TokenData = token.TokenData,
+            UserId = user.UserId.ToString(),
+            ExpiresAt = token.ExpiresAt,
+        };
+
+        return new Response(resp, ContentType.Json);
+    }
+
+    [ApiEndpoint("resetPassword", Method.Post)]
+    [Authentication(false)]
+    public Response ResetPassword(RequestContext context, RealmDatabaseContext database, ApiResetPasswordRequest body)
+    {
+        GameUser? user = database.GetUserFromResetTokenData(body.ResetToken);
+        if (user == null) return new Response(HttpStatusCode.Unauthorized);
+
+        if (body.PasswordSha512.Length != 128)
+            return new Response("Password is definitely not SHA512. Please hash the password - it'll work out better for both of us.",
```

Comment:
You should verify it is SHA512 using a regex here too, not just length

Reply:
I was thinking regex would be overkill for this check, but realistically it should be okay considering it's only a couple hundred more nanoseconds on a whole second

Will be fixed next commit